x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/state_management/stream_enrichment_state_machine/stream_enrichment_state_machine.ts

```diff
@@ -0,0 +1,355 @@
+          reenter: true,
+        },
+      },
+      states: {
+        stream: {
+          initial: 'idle',
+          states: {
+            idle: {
+              on: {
+                'stream.reset': {
+                  guard: 'hasStagedChanges',
+                  target: '#ready',
+                  reenter: true,
+                },
+                'stream.update': {
+                  guard: 'canUpdateStream',
+                  target: 'updating',
+                },
+              },
+            },
+            updating: {
+              invoke: {
+                id: 'upsertStreamActor',
+                src: 'upsertStream',
+                input: ({ context }) => ({
+                  definition: context.definition,
+                  processors: context.processorsRefs
+                    .map((proc) => proc.getSnapshot())
+                    .filter((proc) => proc.matches('configured'))
+                    .map((proc) => proc.context.processor),
+                  fields: context.fields,
+                }),
+                onDone: {
+                  target: 'idle',
+                  actions: [{ type: 'notifyUpsertStreamSuccess' }, { type: 'refreshDefinition' }],
+                },
+                onError: {
+                  target: 'idle',
+                  actions: [{ type: 'notifyUpsertStreamFailure' }],
+                },
+              },
+            },
+          },
+        },
+        enrichment: {
+          type: 'parallel',
+          states: {
+            displayingProcessors: {
+              on: {
+                'processors.add': {
+                  guard: '!hasPendingDraft',
+                  actions: [
+                    { type: 'addProcessor', params: ({ event }) => event },
+                    { type: 'sendProcessorChangeToSimulator' },
+                  ],
+                },
+                'processors.reorder': {
+                  guard: 'hasMultipleProcessors',
+                  actions: [
+                    { type: 'reorderProcessors', params: ({ event }) => event },
+                    { type: 'sendProcessorChangeToSimulator' },
+                  ],
+                },
+                'processor.delete': {
+                  actions: [
+                    { type: 'stopProcessor', params: ({ event }) => event },
+                    { type: 'deleteProcessor', params: ({ event }) => event },
+                    { type: 'sendProcessorChangeToSimulator' },
+                  ],
+                },
+                'processor.change': {
+                  actions: [
+                    { type: 'reassignProcessors' },
+                    { type: 'sendProcessorChangeToSimulator' },
+                  ],
+                },
+              },
+            },
+            displayingSimulation: {
+              entry: [{ type: 'spawnSimulationMachine' }],
+              initial: 'viewDataPreview',
+              states: {
+                viewDataPreview: {
```

Comment:
Is there anything we will actually do with this? Which tab is active seems like UI state to me we don't necessarily need to manage here (another place where we can simplify a bit - we can pull it back into xstate when the active tab actually has a side effect)

Reply:
Xstate is not only for side effects, its strenght is in preventing impossible transitions and describe that state of the UI.

In this case, there are a bunch of events (changes on processors) that should only be available and have effect when the user is specifically previewing the documents, not the detected fields section (I'm adding the specific detected field transitions in the follow-up work). 

This is also specifically used to determine in which state of the simulation we are, if we would just control this in the react component, we'd end up with a bunch of transitions that can be used anywhere, like changing a processor could be invoked in any component.
That would eclissate the benefits of relying on a finite state machine that protect from the impossible scenarios.